### ISSUE TITLE
Endpoints for registering users time out under certain circumstances

### ISSUE THREAD

**Issue by artemislena** [matrix-org/synapse#15364]
### Description

T.: We tried registering a Matrix user on our server. Usually, this useda work by just using the `register_new_matrix_user` script. Yet, regardlessa whether we supplied the registration secret via the config file or directly, n regardlessa whether we supplied the URL ourselves or let't be autodiscovered, 't always hung at `Sending registration request`. So, we tried setting `enable_registration` n `registration_requires_token` to `true`, generated a registration token, n tried registering thru Element (https://element.artemislena.eu) (we dunno whether Element actually supports registration tokens even, n'f not, whether *any* client does, we just tried't out). Well, after entering username n password on the registration window, that hung forever, too. Checked the browser console, apparently we were getting 429 (i.e. rate limit?) on the endpoint for checking whether the username was available (said endpoint simply returned `{"true"}` when we tried opening't manually in the browser, btw). At that point, I asked in the Synapse Admins room, where, after excluding several config issues, I was told trying registering w `synadm`. That timed out too (see logs below) n I was told we should open a GitHub issue, so there we go.

### Steps to reproduce

I dunno really, we were told others (not using workers) weren't experiencing this issue. We didn't experience't in the past, either, so it's gotta be a regression w a recent Synapse version.
- set up a server using our config?
- try registering a user thru anya the methods we tried

### Homeserver

artemislena.eu

### Synapse Version

1.80.0

### Installation Method

Other (please mention below)

### Database

PostgreSQL, single server, no, yes

### Workers

Multiple workers

### Platform

NixOS (Synapse's installed n configured w Nix), RockPro64 4GB (Rockchip RK3399 SoC), 500GB NVMe SSD, running thru systemd (no container besides basic systemd sandboxing, no virtualization)

### Configuration

Message retention's turned on

### Relevant log output

```shell
No ERRORs appeared, at INFO wasn't ablea tell whether anything relevant appeared

synadm logs:

[matrix-synapse@rp64-ssd:~]$ synadm -vv user modify @test_user:artemislena.eu -P 01Elmnj0hLL4rxp7y0j7zbZwPZ78U2fqEs2vnU
DEBUG Config entry read. user: fantasycookie17
DEBUG Config entry read. token: REDACTED
DEBUG Config entry read. base_url: http://localhost:8008
DEBUG Config entry read. admin_path: /_synapse/admin
DEBUG Config entry read. matrix_path: /_matrix
DEBUG Config entry read. timeout: 30
DEBUG Config entry read. server_discovery: well-known
DEBUG Config entry read. homeserver: artemislena.eu
DEBUG Config entry read. format: human
DEBUG Formatter in use: human - <function humanize at 0xe196e4f1c310>
DEBUG A proper MXID was passed.
Current user account settings:
DEBUG A proper MXID was passed.
INFO  Querying get on http://localhost:8008/_synapse/admin/v2/users/@test_user:artemislena.eu
WARNING Synapse returned status code 404
errcode  M_NOT_FOUND
error    User not found
User account settings to be modified:
Password will be set as provided on command line.
Are you sure you want to modify user? (y/N): y
INFO  Querying put on http://localhost:8008/_synapse/admin/v2/users/@test_user:artemislena.eu
ERROR ReadTimeout while querying Synapse: HTTPConnectionPool(host='localhost', port=8008): Read timed out. (read timeout=30)
User could not be modified.
```
```


### Anything else that would be useful to know?

Config at https://codeberg.org/artemislena/nixos-server-configs/src/branch/main/matrix/synapse.nix

**Comment by squahtx** [comment]
Could you post the Synapse logs covering either the `register_new_matrix_user` run, Element Web registration attempt or `synadm` run?

**Comment by artemislena** [comment]
https://pad.artemislena.eu/code/#/2/code/view/-6Fqc8PETkXPMJCMP32GPrXlZR5IAiUQl4REqKj6KA4/present/

L: We uh, ran that exact same synadm line again with Synapse on INFO and uh, grepped the logs for the endpoints it was calling. Apparently, those happened around 15:28:21, so we redacted everything before that except for the startup things, and we also redacted any stuff containing potentially identifying information (like federating servers, room IDs or MXIDs, besides those for our own admin accounts) in between as well as everything up to around 15:28:28; that's what the link above shows. Most relevant lines should be these though, I think?

```
Apr 04 15:28:21 rp64-ssd synapse[26582]: synapse.http.server: [GET-64] <XForwardedForRequest at 0xdf59c47d8af0 method='GET' uri='/_synapse/admin/v2/users/@test_user:artemislena.eu' clientproto='HTTP/1.1' site='8008'> SynapseError: 404 - User not found
Apr 04 15:28:21 rp64-ssd synapse[26582]: synapse.access.http.8008: [GET-64] ::1 - 8008 - {@fantasycookie17:artemislena.eu} Processed request: 0.083sec/0.012sec (0.024sec, 0.003sec) (0.011sec/0.011sec/4) 50B 404 "GET /_synapse/admin/v2/users/@test_user:artemislena.eu HTTP/1.1" "python-requests/2.28.2" [0 dbevts]
```

So, in other words, the PUT it sends never show up in logs on INFO. Which is uh, why we tried on DEBUG too, where all we got was this when grepping:
```
Apr 04 16:05:37 rp64-ssd synapse[29169]: synapse.access.http.8008: [PUT-464] ::1 - 8008 - Received request: PUT /_synapse/admin/v2/users/@test_user:artemislena.eu
Apr 04 16:05:37 rp64-ssd synapse[29169]: synapse.storage.txn: [PUT-464] [TXN START] {is_server_admin-9d6}
Apr 04 16:05:37 rp64-ssd synapse[29169]: synapse.storage.SQL: [PUT-464] [SQL] {is_server_admin-9d6} SELECT admin FROM users WHERE name = ?
Apr 04 16:05:37 rp64-ssd synapse[29169]: synapse.storage.SQL: [PUT-464] [SQL values] {is_server_admin-9d6} ['@fantasycookie17:artemislena.eu']
Apr 04 16:05:37 rp64-ssd synapse[29169]: synapse.storage.SQL: [PUT-464] [SQL time] {is_server_admin-9d6} 0.001414 sec
Apr 04 16:05:37 rp64-ssd synapse[29169]: synapse.storage.txn: [PUT-464] [TXN END] {is_server_admin-9d6} 0.006761 sec
Apr 04 16:05:37 rp64-ssd synapse[29169]: synapse.storage.txn: [_maybe_retry_device_resync-1] [TXN START] {get_user_ids_requiring_device_list_resync-9d7}
Apr 04 16:05:37 rp64-ssd synapse[29169]: synapse.storage.SQL: [_maybe_retry_device_resync-1] [SQL] {get_user_ids_requiring_device_list_resync-9d7} SELECT user_id FROM device_lists_remote_resync
Apr 04 16:05:37 rp64-ssd synapse[29169]: synapse.util.caches.lrucache: [LruCache._expire_old_entries-2] Searching for stale caches
Apr 04 16:05:37 rp64-ssd synapse[29169]: synapse.util.caches.lrucache: [LruCache._expire_old_entries-2] Dropped 0 items from caches
Apr 04 16:05:37 rp64-ssd synapse[29169]: synapse.metrics._gc: [sentinel] Collecting gc 0
Apr 04 16:05:38 rp64-ssd synapse[29169]: synapse.storage.TIME: [sentinel] Total database time: 9.623% {get_received_txn_response(21): 2.269%, set_received_txn_response(21): 1.976%, get_destination_retry_timings(13): 1.874%}
```

Scrolled down a bit further, but uhm, that's all we could find on it (there were a lot of logs per second and for user privacy reasons, we absolutely can't post those unless we're sure they're to do with our requests, so, would need further information on uh, what to search for exactly).

**Comment by reivilibre** [comment]
Please would you be able to write a bit more clearly? The filler words and unusual spellings are making it hard for me to follow this issue. I know it can be a bit dull filling in these issue forms but being able to skim the information is really useful and saves a lot of time on our end, so I'd request your co-operation towards this matter.

In your most recent log snippet, what were you grepping for?
I'd suggest searching for the request ID (in your case: it was `PUT-464` for this request) so we have all the lines for the request (without them being interspersed with lines for other requests).

If you can do that at the DEBUG log level again, that may help.
Which logs correspond to which worker?

I'd suggest trying the request on the main process if that's not already what you're doing.

**Comment by artemislena** [comment]
DeM: Certainly. The requests were done on the main Synapse process, and the logs originate from it. Thank you for pointing out the request ID, that was indeed very helpful. Trying registration with synadm again, we get the following logs:

```
Apr 12 17:39:10 rp64-ssd synapse[10096]: synapse.access.http.8008: [PUT-777] ::1 - 8008 - Received request: PUT /_synapse/admin/v2/users/@test_user:artemislena.eu
Apr 12 17:39:10 rp64-ssd synapse[10096]: synapse.storage.txn: [PUT-777] [TXN START] {is_server_admin-1290}
Apr 12 17:39:10 rp64-ssd synapse[10096]: synapse.storage.SQL: [PUT-777] [SQL] {is_server_admin-1290} SELECT admin FROM users WHERE name = ?
Apr 12 17:39:10 rp64-ssd synapse[10096]: synapse.storage.SQL: [PUT-777] [SQL values] {is_server_admin-1290} ['@fantasycookie17:artemislena.eu']
Apr 12 17:39:10 rp64-ssd synapse[10096]: synapse.storage.SQL: [PUT-777] [SQL time] {is_server_admin-1290} 0.003218 sec
Apr 12 17:39:10 rp64-ssd synapse[10096]: synapse.storage.txn: [PUT-777] [TXN END] {is_server_admin-1290} 0.007602 sec
```

After a few seconds, these disappear (at DEBUG, Synapse appears to generate so many logs that our journal fills up eventually), and after a few more seconds (at the point where synadm reports a timeout), this one becomes visible: `Apr 12 17:39:40 rp64-ssd synapse[10096]: synapse.http.site: [PUT-777] Connection from client lost before response was sent`

**Comment by reivilibre** [comment]
Is it possible you're missing logs in between because they're being rotated so quickly?
If you don't want to mess with the journal size limit but you have enough disk space to get ~60 seconds of logs, I would be tempted to start `journalctl -fu synapse > tmp`, make the request, wait ~60 sec, then interrupt it and then grep through that file.

Registration isn't a particularly complicated procedure (the PUT endpoint used by synadm a little bit moreso, but if even `register_new_matrix_user` is timing out then it's very confusing), so it's strange that this is timing out.

I am tempted to suggest patching in some extra logging, I have some notes for doing that on NixOS somewhere.

**Comment by artemislena** [comment]
DeM: To rule out that possibility, we tried again using `journalctl -f` to save things into a file, as you suggested. Increasing the timeout of synadm from 30 to 120 resulted in no change in logs (apart from time intervals between the first couple lines and the last line). We did also try registering again with the `register_new_matrix_user` script. Grepping the logs for the endpoint it uses only yielded
```
Apr 12 18:44:56 rp64-ssd synapse[14123]: synapse.access.http.8008: [GET-8869] ::1 - 8008 - Received request: GET /_synapse/admin/v1/register
Apr 12 18:44:56 rp64-ssd synapse[14123]: synapse.access.http.8008: [GET-8869] ::1 - 8008 - {None} Processed request: 0.002sec/0.077sec (0.002sec, 0.000sec) (0.000sec/0.000sec/0) 140B 200 "GET /_synapse/admin/v1/register HTTP/1.1" "python-requests/2.28.2" [0 dbevts]
Apr 12 18:44:56 rp64-ssd synapse[14123]: synapse.access.http.8008: [POST-8872] ::1 - 8008 - Received request: POST /_synapse/admin/v1/register
```

Grepping for the respective request IDs did not find further results except for these lines.

**Comment by artemislena** [comment]
DeM: Well, for the log outputs, we did try registering `test_user` every time, but we originally noticed it while trying to register a different username. I am afraid we cannot send you any logs that are not clearly related to the registration process ‒ there are different people outside our system using our server, and we never specified a privacy policy to them that would allow for such.

**Comment by artemislena** [comment]
T.: We figured out the issue; our `bcrypt_rounds` were too high. We had been using `ssh-keygen` as a benchmark for increasing, since we read that used sth called `bcrypt_pbkdf`, but I'm guessing that's quite different from plain bcrypt after all (we couldn't find any docs on't so originally guessed't'd just mix bcrypt n PBKDF(2))? `mkpasswd -m bcrypt` 's a much better benchmark n we figured the default work factora 12 was just fine. Feel free reopening this w a new name (sth like missing logging on password hash derivation or sth), or, even better, just making a new issue, but I'm closing this for now. (tbf, the docs mention setting'ta 24 might take >20 minutes, I think we must've just overread the sentence; we set'ta 30 cause `ssh-keygen -a 30` had no noticeable delay after entering a passphrase, so people reading the docs properly shouldn't ever run into this issue) Sorry for all this trouble.
